File: packages/storage/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/storage/FirebaseStoragePlugin.java
package io.capawesome.capacitorjs.plugins.firebase.storage;

import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.firebase.storage.classes.options.DeleteFileOptions;
import io.capawesome.capacitorjs.plugins.firebase.storage.classes.options.GetDownloadUrlOptions;
import io.capawesome.capacitorjs.plugins.firebase.storage.classes.options.GetMetadataOptions;
import io.capawesome.capacitorjs.plugins.firebase.storage.classes.options.ListFilesOptions;
import io.capawesome.capacitorjs.plugins.firebase.storage.classes.options.UpdateMetadataOptions;
import io.capawesome.capacitorjs.plugins.firebase.storage.classes.options.UploadFileOptions;
import io.capawesome.capacitorjs.plugins.firebase.storage.interfaces.EmptyResultCallback;
import io.capawesome.capacitorjs.plugins.firebase.storage.interfaces.NonEmptyResultCallback;
import io.capawesome.capacitorjs.plugins.firebase.storage.interfaces.Result;

@CapacitorPlugin(name = "FirebaseStorage")
public class FirebaseStoragePlugin extends Plugin {

    public static final String TAG = "FirebaseStorage";
    public static final String ERROR_PATH_MISSING = "path must be provided.";
    public static final String ERROR_URI_MISSING = "uri must be provided.";

    private FirebaseStorage implementation;

    public void load() {
        implementation = new FirebaseStorage(this);
    }

    @PluginMethod
    public void deleteFile(PluginCall call) {
        try {
            String path = call.getString("path");
            if (path == null) {
                call.reject(ERROR_PATH_MISSING);
                return;
            }

            DeleteFileOptions options = new DeleteFileOptions(path);
            EmptyResultCallback callback = new EmptyResultCallback() {
                @Override
                public void success() {
                    call.resolve();
                }

                @Override
                public void error(Exception exception) {
                    Logger.error(TAG, exception.getMessage(), exception);
                    call.reject(exception.getMessage());
                }
            };

            implementation.deleteFile(options, callback);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void getDownloadUrl(PluginCall call) {
        try {
            String path = call.getString("path");
            if (path == null) {
                call.reject(ERROR_PATH_MISSING);
                return;
            }

            GetDownloadUrlOptions options = new GetDownloadUrlOptions(path);
            NonEmptyResultCallback callback = new NonEmptyResultCallback() {
                @Override
                public void success(Result result) {
                    call.resolve(result.toJSObject());
                }

                @Override
                public void error(Exception exception) {
                    Logger.error(TAG, exception.getMessage(), exception);
                    call.reject(exception.getMessage());
                }
            };

            implementation.getDownloadUrl(options, callback);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void getMetadata(PluginCall call) {
        try {
            String path = call.getString("path");
            if (path == null) {
                call.reject(ERROR_PATH_MISSING);
                return;
            }

            GetMetadataOptions options = new GetMetadataOptions(path);
            NonEmptyResultCallback callback = new NonEmptyResultCallback() {
                @Override
                public void success(Result result) {
                    call.resolve(result.toJSObject());
                }

                @Override
                public void error(Exception exception) {
                    Logger.error(TAG, exception.getMessage(), exception);
                    call.reject(exception.getMessage());
                }
            };

            implementation.getMetadata(options, callback);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void listFiles(PluginCall call) {
        try {
            String path = call.getString("path");
            if (path == null) {
                call.reject(ERROR_PATH_MISSING);
                return;
            }
            int maxResults = call.getInt("maxResults", 1000);
            String pageToken = call.getString("pageToken");

            ListFilesOptions options = new ListFilesOptions(path, maxResults, pageToken);
            NonEmptyResultCallback callback = new NonEmptyResultCallback() {
                @Override
                public void success(Result result) {
                    call.resolve(result.toJSObject());
                }

                @Override
                public void error(Exception exception) {
                    Logger.error(TAG, exception.getMessage(), exception);
                    call.reject(exception.getMessage());
                }
            };

            implementation.listFiles(options, callback);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void updateMetadata(PluginCall call) {
        try {
            String path = call.getString("path");
            if (path == null) {
                call.reject(ERROR_PATH_MISSING);
                return;
            }
            JSObject metadata = call.getObject("metadata");

            UpdateMetadataOptions options = new UpdateMetadataOptions(path, metadata);
            EmptyResultCallback callback = new EmptyResultCallback() {
                @Override
                public void success() {
                    call.resolve();
                }

                @Override
                public void error(Exception exception) {
                    Logger.error(TAG, exception.getMessage(), exception);
                    call.reject(exception.getMessage());
                }
            };

            implementation.updateMetadata(options, callback);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod(returnType = PluginMethod.RETURN_CALLBACK)
    public void uploadFile(PluginCall call) {
        try {
            String path = call.getString("path");
            if (path == null) {
                call.reject(ERROR_PATH_MISSING);
                return;
            }
            String uri = call.getString("uri");
            if (uri == null) {
                call.reject(ERROR_URI_MISSING);
                return;
            }
            String callbackId = call.getCallbackId();

            UploadFileOptions options = new UploadFileOptions(path, uri, callbackId);
            NonEmptyResultCallback callback = new NonEmptyResultCallback() {
                @Override
                public void success(Result result) {
                    call.resolve(result.toJSObject());
                }

                @Override
                public void error(Exception exception) {
                    Logger.error(TAG, exception.getMessage(), exception);
                    call.reject(exception.getMessage());
                }
            };

            implementation.uploadFile(options, callback);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }
}

File: packages/storage/ios/Plugin/FirebaseStoragePlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(FirebaseStoragePlugin)
public class FirebaseStoragePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "FirebaseStoragePlugin"
    public let jsName = "FirebaseStorage"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "deleteFile", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getDownloadUrl", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getMetadata", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "listFiles", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "updateMetadata", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "uploadFile", returnType: CAPPluginReturnCallback),
        CAPPluginMethod(name: "useEmulator", returnType: CAPPluginReturnPromise)
    ]
    public let tag = "FirebaseFirestore"
    public let errorPathMissing = "path must be provided."
    public let errorUriMissing = "uri must be provided."
    public let errorMetadataMissing = "metadata must be provided."
    public let errorCallbackIdMissing = "callbackId must be provided."
    public let errorFileNotExist = "File does not exist."
    public let errorHostMissing = "host must be provided."
    private var implementation: FirebaseStorage?

    override public func load() {
        self.implementation = FirebaseStorage(plugin: self)
    }

    @objc func deleteFile(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject(errorPathMissing)
            return
        }

        let options = DeleteFileOptions(path: path)

        implementation?.deleteFile(options, completion: { error in
            if let error = error {
                CAPLog.print("[", self.tag, "] ", error)
                call.reject(error.localizedDescription, FirebaseStorageHelper.createErrorCode(error: error))
                return
            }
            call.resolve()
        })
    }

    @objc func getDownloadUrl(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject(errorPathMissing)
            return
        }

        let options = GetDownloadUrlOptions(path: path)

        implementation?.getDownloadUrl(options, completion: { result, error in
            if let error = error {
                CAPLog.print("[", self.tag, "] ", error)
                call.reject(error.localizedDescription, FirebaseStorageHelper.createErrorCode(error: error))
                return
            }
            if let result = result?.toJSObject() as? JSObject {
                call.resolve(result)
            }
        })
    }

    @objc func getMetadata(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject(errorPathMissing)
            return
        }

        let options = GetMetadataOptions(path: path)

        implementation?.getMetadata(options, completion: { result, error in
            if let error = error {
                CAPLog.print("[", self.tag, "] ", error)
                call.reject(error.localizedDescription, FirebaseStorageHelper.createErrorCode(error: error))
                return
            }
            if let result = result?.toJSObject() as? JSObject {
                call.resolve(result)
            }
        })
    }

    @objc func listFiles(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject(errorPathMissing)
            return
        }
        let maxResults = call.getInt("maxResults") ?? 1000
        let pageToken = call.getString("pageToken")

        let options = ListFilesOptions(path: path, maxResults: maxResults, pageToken: pageToken)

        implementation?.listFiles(options, completion: { result, error in
            if let error = error {
                CAPLog.print("[", self.tag, "] ", error)
                call.reject(error.localizedDescription, FirebaseStorageHelper.createErrorCode(error: error))
                return
            }
            if let result = result?.toJSObject() as? JSObject {
                call.resolve(result)
            }
        })
    }

    @objc func updateMetadata(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject(errorPathMissing)
            return
        }
        guard let metadata = call.getObject("metadata") else {
            call.reject(errorMetadataMissing)
            return
        }

        let options = UpdateMetadataOptions(path: path, metadata: metadata)

        implementation?.updateMetadata(options, completion: { error in
            if let error = error {
                CAPLog.print("[", self.tag, "] ", error)
                call.reject(error.localizedDescription, FirebaseStorageHelper.createErrorCode(error: error))
                return
            }
            call.resolve()
        })
    }

    @objc func uploadFile(_ call: CAPPluginCall) {
        call.keepAlive = true

        guard let path = call.getString("path") else {
            call.reject(errorPathMissing)
            return
        }
        guard let uri = call.getString("uri") else {
            call.reject(errorUriMissing)
            return
        }
        guard let url = URL(string: uri) else {
            call.reject(errorFileNotExist)
            return
        }
        guard let callbackId = call.callbackId else {
            call.reject(errorCallbackIdMissing)
            return
        }
        let metadata = call.getObject("metadata")

        let options = UploadFileOptions(path: path, uri: url, metadata: metadata, callbackId: callbackId)

        implementation?.uploadFile(options, completion: { result, error, releaseCall in
            if let error = error {
                CAPLog.print("[", self.tag, "] ", error)
                call.reject(error.localizedDescription, FirebaseStorageHelper.createErrorCode(error: error))
                return
            }
            if let result = result?.toJSObject() as? JSObject {
                call.resolve(result)
            }
            if releaseCall == true {
                self.bridge?.releaseCall(call)
            }
        })
    }

    @objc func useEmulator(_ call: CAPPluginCall) {
        guard let host = call.getString("host") else {
            call.reject(errorHostMissing)
            return
        }
        let port = call.getInt("port") ?? 9199

        implementation?.useEmulator(host, port)
        call.resolve()
    }
}
